File: Leetcode/reverse_linkedlist.py
# https://leetcode.com/problems/reverse-linked-list/
'''
    Author - Subhajit Das
    University of Engineering and Management, Kolkata
    21/10/2019
'''


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if(head == None or head.next == None):
            return head

        returnRoot = self.reverseList(head.next)

        head.next.next = head
        head.next = None
        return returnRoot

File: Leetcode/reverse_linkedlist.java
// https://leetcode.com/problems/reverse-linked-list/
/*
    Author @Subhajit Das (sdsubhajitdas.github.io)
    SWE @Turbot HQ India PVT Ltd.
    21/11/2020
*/


// Definition for singly-linked list.
// class ListNode {
//     int val;
//     ListNode next;
//     ListNode() {}
//     ListNode(int val) { this.val = val; }
//     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
// }

class Solution {
    public ListNode reverseList(ListNode head) {

        if(head == null || head.next == null) return head;

        ListNode previousNode=null, nextNode;

        while(head!=null){
            nextNode = head.next;
            head.next = previousNode;
            previousNode = head;
            head = nextNode;
        }
        return previousNode;
    }
}


